ZOOKEEPER-2122: add SSL support for C-client

symat opened the pull request:
This PR is based on the works of Asnish Amarnath and Suhas Dantkale. Most of the kudos should go to them and those who were reviewing all the previous PRs.

**The PR includes the following changes from PR#639:**
- OPENSSL 1.1.1 version support in C-client

**The PR includes the following changes from PR#990:**
- also supporting OPENSSL 1.0.2
- SSL connection on non-blocking socket is handled correctly
- Support of Certificate Chains
- Fix Memory leaks
- Dynamically generated test certificates

**The following new changes were added into the PR:**
- fix CMake + VisualStudio2019 build with windows
- fix C CLI to compile / work both with windows and linux (I tested them manually)
- fix (and simplify) the way how the server is started with C unit tests, so it is compatible with maven build now
- the test case `testReadOnly` was failing with the previous PR because there was a bug in the C-client code, I fixed that
- I also added new test case: `testReadOnlyWithSSL`

**Testing this PR on linux:**
```
git clean -xdf

# compile ZooKeeper server plus the C-client code
mvn clean install -DskipTests -Pfull-build

# compile and execute C-client unit tests
cd zookeeper-client/
mvn clean install -Pfull-build
```


**Compile the code on windows (only cmake is supported):**
- download C-Make:  https://cmake.org/download/
- Install community edition of Visual Studio: https://visualstudio.microsoft.com/downloads/
- Download OpenSSL (e.g. 1.0.2): https://slproweb.com/products/Win32OpenSSL.html (e.g. install it to `c:\OpenSSL-Win64`)
- compile the java code using: `mvn clean install -DskipTests`
- go to the Client folder: `cd zookeeper-client\zookeeper-client-c`
- configure the project:  `cmake . -D WITH_OPENSSL=c:\OpenSSL-Win64`
- build the project: `cmake --build .`

**Testing the C-client with SSL manually:**
- run the `zookeeper-client/zookeeper-client-c/ssl/gencerts.sh` to generate certificate files (e.g. copy it to an empty folder like `/tmp/ssl/` and start is)
- start a ZooKeeper server, using some config file like this one:
```
tickTime=3000
initLimit=10
syncLimit=5
dataDir=/tmp/zkdata

secureClientPort=22281
serverCnxnFactory=org.apache.zookeeper.server.NettyServerCnxnFactory
ssl.keyStore.location=/tmp/ssl/server.jks
ssl.keyStore.password=password
ssl.trustStore.location=/tmp/ssl/servertrust.jks
ssl.trustStore.password=password
```
- start the command line client (cli.exe on windows, cli_mt or cli_st on linux): `./cli_mt --host localhost:22281 --ssl /tmp/ssl/server.crt,/tmp/ssl/client.crt,/tmp/ssl/clientkey.pem,password`

suhasdantkale commented on zookeeper-client/zookeeper-client-c/tests/TestReadOnlyClient.cc:
```diff
@@ -81,29 +85,68 @@ class Zookeeper_readOnly : public CPPUNIT_NS::TestFixture {
+        return zh;
+    }
+
+    void testReadOnlyWithSSL() {
```
Good to see this test being added.

symat commented:
retest maven build

symat commented:
retest maven build

anmolnar commented on zookeeper-client/zookeeper-client-c/ssl/gencerts.sh:
```diff
@@ -0,0 +1,118 @@
+ST = California
+L = San Francisco
+O = Bookkeeper
+emailAddress = dev@bookkeeper.apache.org
```
Shall we change this to `zookeeper.apache.org`? :)

suhasdantkale replied:
+1. Sorry. That was my mistake.

symat replied:
I will change it

suhasdantkale commented:
@symat 
Overall the changes look good to me.
For the Windows side of the changes, since  I don't have any experience, I can't comment anything there.

anmolnar commented:
@symat Found 3 issues:

- `gencerts.sh` should use FQDN instead of `zookeeper.apache.org` as CN in order to make local testing easier (not really an issue),
- the CLI command line doesn't need `--host` as a parameter, only the list of hosts,
- patch doesn't work for me:

I specified the stores for both client and quorum communication: quorum was successfully brought up with TLS enabled, but client is unable to connect. Error message in server log:

```
2019-11-18 15:52:43,738 [myid:1] - ERROR [nioEventLoopGroup-4-6:NettyServerCnxnFactory$CertificateVerifier@386] - Unsuccessful handshake with session 0x0
2019-11-18 15:52:43,738 [myid:1] - WARN  [nioEventLoopGroup-4-6:NettyServerCnxnFactory$CnxnChannelHandler@228] - Exception caught
io.netty.handler.codec.DecoderException: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record: 0000002d000000000000000000000000000075300000000000000000000000100000000000000000000000000000000000
```

We can take a look tomorrow in person.

anmolnar commented:
Retested the patch with @symat in person and issue 2. and 3. are already resolved.
Please change gencerts script and fix documentation and we're done.

symat commented:
@anmolnar thanks for testing, I changed the the SSL certificate generator and updated the `readme_packaging.md` mentioning the new OpenSSL library dependency.

symat commented:
retest maven build

ztzg commented:
I experienced [a failed build](https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/1633/) with this error:

     configure: error: --with-openssl path does not point to a directory
     [ERROR] Command execution failed.

It seems that this stanza (in `pom.xml`) is not valid on all build hosts:

    <argument>--with-openssl=/usr/include/openssl/</argument>

Not sure how we can make sure that OpenSSL headers are installed on all Jenkins workers?

It might also be nice to use autodetection, at least for `autoconf`; here is what I ended up with in v2 of the SASL patches (comments welcome; I'm trying to "align" the two features):

https://github.com/apache/zookeeper/blob/f248c53089d82d2372404f0a7ca992bebd6ecf68/zookeeper-client/zookeeper-client-c/configure.ac#L114-L142

anmolnar commented:
Thanks @ztzg ! Great catch.
@symat Would you please create another PR with the fix?

symat commented:
@ztzg thanks for the notice and the proposed fix! I think your approach is great, I will copy it :)

Maybe I will also include a maven parameter, so that you can also specify the OpenSSL header files location also when using `mvn` and not building with `make`.

symat commented:
@ztzg did you also implement the same autodetection logics for SASL in cmake?
I would be happy to copy your cmake approach as well if there is any :)

I am not sure which way is the default one... right now maven is using autotools / make, but e.g. on windows only CMake is working / supported AFAIK. Did you test SASL on windows? (personally I wouldn't be too sad if it wouldn't work, but then maybe it makes sense to add a Jira ticket so implement it)

ztzg commented:
Hi @symat,

> @ztzg thanks for the notice and the proposed fix! I think your approach is great, I will copy it :)

You're welcome!  (It's not like it's original, anyway; I stole bits from somewhere else :)

> Maybe I will also include a maven parameter, so that you can also specify the OpenSSL header files location also when using `mvn` and  not building with `make`.

Good idea.  I will align my patches.

> @ztzg did you also implement the same autodetection logics for SASL in cmake? I would be happy to copy your cmake approach as well if there is any :)

No, I haven't looked into that yet, sorry.  The "CMake way" usually involves "modules," and it looks like there is a "built-in" one for OpenSSL:

https://cmake.org/cmake/help/v3.16/module/FindOpenSSL.html
https://stackoverflow.com/a/45548831

> I am not sure which way is the default one... right now maven is using autotools / make, but e.g. on windows only CMake is working / supported AFAIK.

My understanding is that CMake is there for facilitating Windows builds, and that there are no plans to remove Autoconf at this point.  I may be wrong, though.

> Did you test SASL on windows? (personally I wouldn't be too sad if it wouldn't work, but then maybe it makes sense to add a Jira ticket so implement it)

I haven't, yet.  As far as I can tell, Cyrus SASL isn't officially supported on Windows; the ["official" project page](https://www.cyrusimap.org/sasl/sasl/windows.html) says:

>> Note, that Cyrus SASL on Windows is still laregely a “work in progress”

So I'll have to figure out how to build it at all before integrating with CMake :)

symat commented:
@ztzg thanks for the reply / help!

I created ZOOKEEPER-3630, hopefully I can start working on it later this week (the autotools part is quite easy, but the windows / CMake part is always tricky, as I have to setup a VM for testing)

kaosmonk commented:
> @symat Found 3 issues:
> 
> * `gencerts.sh` should use FQDN instead of `zookeeper.apache.org` as CN in order to make local testing easier (not really an issue),
> * the CLI command line doesn't need `--host` as a parameter, only the list of hosts,
> * patch doesn't work for me:
> 
> I specified the stores for both client and quorum communication: quorum was successfully brought up with TLS enabled, but client is unable to connect. Error message in server log:
> 
> ```
> 2019-11-18 15:52:43,738 [myid:1] - ERROR [nioEventLoopGroup-4-6:NettyServerCnxnFactory$CertificateVerifier@386] - Unsuccessful handshake with session 0x0
> 2019-11-18 15:52:43,738 [myid:1] - WARN  [nioEventLoopGroup-4-6:NettyServerCnxnFactory$CnxnChannelHandler@228] - Exception caught
> io.netty.handler.codec.DecoderException: io.netty.handler.ssl.NotSslRecordException: not an SSL/TLS record: 0000002d000000000000000000000000000075300000000000000000000000100000000000000000000000000000000000
> ```
> 
> We can take a look tomorrow in person.

Apologies for using this thread, but I am experiencing this very same error as above in my 3 node cluster (3.5.5). What was the solution here?
